### DISCUSSION TITLE
throwing error inside `onError` callback gets ignored

### DISCUSSION THREAD

**Discussion by kshanmugharaj** [TanStack/query#3862] [Q&A] [answered]
Currently I'm using `v3.x.x` I have a code like this (roughly)

```
const useUpdateTodo = () => {
  return useMutation((data: any) => {
    // to simulate error case
    throw data;
    // return axios.put("https://jsonplaceholder.typicode.com/posts/1", data);
  }, {
    useErrorBoundary: true
  });
};
```

then in component I use like 

```
  const handleSave = (returnPromise = false) => {
    todoMutation.mutate(
      {
        id: 1,
        title: "foo",
        body: "bar",
        userId: 1,
      },
      {
        onSuccess: (res) => {
          console.log(res);
        },
        onError: (e) => {
          console.log(e);

          // if (some condition) {
          //   show toast notification.
          // }

          throw e; // This  should get logged by logger which listens to unhandled exception.
        },
      }
    );
  };
```

But `throw e` is somehow absorbed and the logger is not logging the error.

Once I changed the  `throw e; ` to `return Promise.reject(err)` then it's working as expected.

In `v4` `throw e` is working. Just wanted to know is this bug in `v3` or it's intended behaviour?

**Comment by TkDodo** [comment]
not sure, which v3 do you have?

**Reply by kshanmugharaj** [reply]
`"react-query": "^3.9.8"`

**Reply by TkDodo** [reply]
can you try with the latest v3?

**Reply by kshanmugharaj** [reply]
Yes, `v4.0.0` is working, I mean it's throwing error. But sorry one more question `setLogger` if I import like

```
import { QueryClient, setLogger } from "react-query";
```

it's giving 

```
Uncaught SyntaxError: The requested module '/node_modules/.vite/deps/react-query.js?v=5e8c9226' does not provide an export named 'setLogger' (at query-client.ts:2:1)
```

**Reply by TkDodo** [reply]
setLogger has been removed in v4, please see the migration guide.

**Reply by kshanmugharaj** [reply]
Sure, am I missing something https://react-query-v3.tanstack.com/installation this I installed. That link has `setLogger ` in it.

**Reply by kshanmugharaj** [reply]
BTW, Is it fine to throw error like that in `react-query` perspective?

```
onError: (e) => {
          console.log(e);

          // if (some condition) {
          //   show toast notification.
          // }

          throw e; // This  should get logged by logger which listens to unhandled exception.
        },
```

**Reply by kshanmugharaj** [reply]
> not sure, which v3 do you have?

Updating to `v4.0.0` is working as expected.

**Comment by kshanmugharaj** [comment, marked as answer]
In `v4.0.0` it behaving as expected as @TkDodo suggested.
